mantidproject/mantid
# Allow cmake to configure for a docs-only or workbench-only build

**Description of work.**

Both the documentation and workbench can only be built with one of the other options such as framework or mantidqt enabled. To allow for flexibility, if both mantidqt & framework are importable already they we should be able to build & install the documentation or workbench.

This work allows this and is a precursor to creating a new, separate noarch `mantiddocs` Conda package. The current `mantidworkbench` package contains the documentation and has to be built 3 times. By moving out the documentation it can be built once on the fastest node and simply included by other packages.



**To test:**

The CI builds will check the current incremental configuration.

*Note: We do not want to use the standard mantid-developer environment here. Don't activate that.*

1 - Create a clean build directory for testing only workbench:
* Change to the directory on the command line
* Create a test conda environment and install `mantidqt`: `mamba create -p ./condaenv -c mantid/label/nightly mantidqt cmake ninja`
* Activate it: 
  * macOS/Linux: `mamba activate $PWD/condaenv`
  * Windows: `mamba activate %CD%\condaenv`
* Configure CMake:
  * macOS/Linux: `cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -DCONDA_ENV=TRUE -DMANTID_FRAMEWORK_LIB=SYSTEM -DMANTID_QT_LIB=SYSTEM -DENABLE_WORKBENCH=ON -DENABLE_DOCS=OFF -B . -S ..`
  * Windows: `cmake -G "Visual Studio 16 2019" -DCONDA_ENV=TRUE -DMANTID_FRAMEWORK_LIB=SYSTEM -DMANTID_QT_LIB=SYSTEM -DENABLE_WORKBENCH=ON -DENABLE_DOCS=OFF -B . -S ..`
* Build:
  * macOS/Linux: `cmake --build .` 
  * Windows: `cmake --build . --config Release`
* Check `bin` (macOS/Linux) or `bin\Release` (Windows) see that workbench exists but not the other libraries

2 - Create a clean build directory for testing only docs:
* Change to the directory on the command line
* Create a test conda environment and install `mantidqt`: `mamba create -p ./condaenv -c mantid/label/nightly mantidqt cmake ninja`
* Activate it: 
  * macOS/Linux: `mamba activate $PWD/condaenv`
  * Windows: `mamba activate %CD%\condaenv`
* Configure CMake:
  * macOS/Linux: `cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -DMANTID_FRAMEWORK_LIB=SYSTEM -DMANTID_QT_LIB=SYSTEM -DENABLE_WORKBENCH=OFF -DENABLE_DOCS=ON -DDOCS_MATH_EXT=sphinx.ext.mathjax -DDOCS_PLOTDIRECTIVE=OFF -DDOCS_DOTDIAGRAMS=OFF -DDOCS_SCREENSHOTS=OFF`
  * Windows: `cmake -G "Visual Studio 16 2019" -DMANTID_FRAMEWORK_LIB=SYSTEM -DMANTID_QT_LIB=SYSTEM -DENABLE_WORKBENCH=OFF -DENABLE_DOCS=ON -DDOCS_MATH_EXT=sphinx.ext.mathjax -DDOCS_PLOTDIRECTIVE=OFF -DDOCS_DOTDIAGRAMS=OFF -DDOCS_SCREENSHOTS=OFF`
* Build:
  * macOS/Linux: `cmake --build . --target docs-qthelp` 
  * Windows: `cmake --build . --config Release --target docs-qthelp`
* Check `bin` (macOS/Linux) or `bin\Release` (Windows) and see they are clean
* Check `docs/qthelp` to see the documentation
 

*There is no associated issue.*

*This does not require release notes* because **these changes are internal and not visible to users.**



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
